### PR TITLE
Add 3.5.1-SNAPSHOT Shim

### LINKED ISSUE #9258: [FEA] Create spark 3.5.1 shim and build env
**Is your feature request related to a problem? Please describe.**
Spark 3.5.0 is released, we need to add 3.5.1 snapshot shim and build env

The following issues are related to this Epic 

- [x] https://github.com/NVIDIA/spark-rapids/issues/9875
- [x] https://github.com/NVIDIA/spark-rapids/issues/9859
- [x] https://github.com/NVIDIA/spark-rapids/issues/9743
- [x] https://github.com/NVIDIA/spark-rapids/issues/9472
- [x] https://github.com/NVIDIA/spark-rapids/issues/10028

### LINKED ISSUE #9743: [BUG][AUDIT] SPARK-45652 - SPJ: Handle empty input partitions after dynamic filtering
**Describe the bug**
https://issues.apache.org/jira/browse/SPARK-45652 was part of feature for storage partitioned join. It changes BatchScanExec which we have the equivalent code in GpuBatchScanExec so we should pull this change in.

https://github.com/apache/spark/commit/75aed566018

### LINKED ISSUE #9859: [AUDIT] [SPARK-45786] Inaccurate Decimal multiplication and division results
**Describe the bug**
https://issues.apache.org/jira/browse/SPARK-45786 fixes inaccurate decimal multiplication and division in Spark. This looks like a dup of https://issues.apache.org/jira/browse/SPARK-40129.

I believe we explicitly copied the behavior of Spark for the multiplication side so we should look to see if we need to change it.

### LINKED ISSUE #9875: [AUDIT] SPARK-43393 Sequence expression can overflow
[SPARK-43393](https://issues.apache.org/jira/browse/SPARK-43393) updated the calculation of the sequence length to detect overflows at https://github.com/apache/spark/commit/b7bc127340c.  We have translated the original approach to a GPU algorithm and may need to update it in a similar way to detect the overflow case.

### PR DESCRIPTION
**Pull request by razajafri**
This PR adds shims for Spark 3.5.1-SNAPSHOT. 

**Changes Made:**
* The following `Shimplify` command was run
```
mvn generate-sources -Dshimplify=true -Dshimplify.move=true -Dshimplify.overwrite=true -Dshimplify.add.shim=351 -Dshimplify.add.base=350
```
The only files that were manually changed were `pom.xml` and `ShimServiceProvider.scala` to add the SNAPSHOT version to the `VERSIONNAMES`. Also removed some empty lines as a result of the above `Shimplify` command

* Added a `DecimalUtilShims.scala` which calls the respective multiplication method depending on the Spark version. In Spark 3.5.1 and other versions, the multiplication doesn't perform an interim cast and as part of spark-rapids-jni [PR](https://github.com/NVIDIA/spark-rapids-jni/pull/1623) another method called `mul128` was added which skips the interim cast. 
* Added `ComputeSequenceSize.scala` to provide a shim for the new method to calculate sequence size and to make sure it's within limit. 
* Made relevant changes to `GpuBatchScanExec` to match the changes in Spark 

**Tests:**
All integration tests were run locally

fixes #9258 
fixes #9859 
fixes #9875
fixes #9743

### CODE REVIEW

**Review by jlowe — APPROVED** [review]

**Review by andygrove — DISMISSED** [review]

**Comment by tgravescs** [comment]
I'm assuming your decimal multiple is related to https://github.com/NVIDIA/spark-rapids/issues/9859...  If so pleas emake sure it fixes it all the way or we comment on that issue.  the shim is very hard to read, one calls mul128 the other calls multiply128. I haven't went and looked at those but one its hard to even see that diff so you should in the very least add a comment or point to issue and explain.

**Comment by razajafri** [comment]
> the shim is very hard to read, one calls mul128 the other calls multiply128. I haven't went and looked at those but one its hard to even see that diff so you should in the very least add a comment or point to issue and explain.

I will go ahead and put in some comments to highlight the change

**Comment by razajafri** [comment]
> I'm assuming your decimal multiple is related to https://github.com/NVIDIA/spark-rapids/issues/9859... If so pleas emake sure it fixes it all the way or we comment on that issue

Discussed this offline. I missed the division bit of the puzzle. Will verify division and post an update here

**Comment by razajafri** [comment]
> > I'm assuming your decimal multiple is related to #9859... If so pleas emake sure it fixes it all the way or we comment on that issue
> 
> Discussed this offline. I missed the division bit of the puzzle. Will verify division and post an update here

I have verified the Decimal division and we match Spark 3.5.1 output. 

It turns out that we were always doing the right thing on the GPU for decimal division. So to match Spark bug for bug we should "fix" the versions Databricks 330+ and Spark versions 340+ by returning the bad answer. I have created an issue for that [here](https://github.com/NVIDIA/spark-rapids/issues/9998)

**Comment by razajafri** [comment]
build

**Comment by razajafri** [comment]
premerge failing due to an unrelated change

```
[2023-12-18T20:02:45.179Z] [ERROR] /home/ubuntu/spark-rapids/tests/src/test/scala/org/apache/spark/sql/rapids/metrics/source/MockTaskContext.scala:69: overriding method getKillReason in class TaskContext of type ()Option[String];
[2023-12-18T20:02:45.180Z]  method getKillReason has weaker access privileges; it should be public
[2023-12-18T20:02:45.180Z] [ERROR]   override private[spark] def getKillReason() = None
[2023-12-18T20:02:45.180Z] [ERROR]                               ^
[2023-12-18T20:02:45.180Z] [ERROR] one error found
```

**Comment by razajafri** [comment]
@NvTimLiu @pxLi why is the premerge picking up databricks 13.3

**Comment by razajafri** [comment]
build

**Comment by razajafri** [comment]
build

**Comment by razajafri** [comment]
> @NvTimLiu @pxLi why is the premerge picking up databricks 13.3

Nevermind, I saw that there is a function that checks if there are any changes in the Databricks shims in addition to the word `databricks` in the title of the PR

**Comment by razajafri** [comment]
build

**Comment by razajafri** [comment]
build

**Comment by razajafri** [comment]
Connectivity problem 
```
...
Could not transfer artifact org.apache.hadoop:hadoop-project-dist:pom:3.1.1.7.1.7.1000-141 from/to cloudera-repo (https://repository.cloudera.com/artifactory/cloudera-repos/): Transfer failed for https://repository.cloudera.com/artifactory/cloudera-repos/org/apache/hadoop/hadoop-project-dist/3.1.1.7.1.7.1000-141/hadoop-project-dist-3.1.1.7.1.7.1000-141.pom 500 Internal Server Error -> [Help 1]
```

**Comment by razajafri** [comment]
build

**Comment by razajafri** [comment]
build

**Comment by gerashegalov** [comment]
build

**Comment by razajafri** [comment]
build

**Comment by razajafri** [comment]
build

**Comment by razajafri** [comment]
I have reverted the tests for versions that we don't support yet. They will be added in other shims

**Comment by razajafri** [comment]
build

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/arithmetic.scala [hunk 1]
```diff
@@ -53,9 +54,8 @@ object AddOverflowChecks {
     withResource(signDiffCV) { signDiff =>
       withResource(signDiff.any()) { any =>
         if (any.isValid && any.getBoolean) {
-          throw RapidsErrorUtils.arithmeticOverflowError(
-            "One or more rows overflow for Add operation."
-          )
+          throw RapidsErrorUtils.
+            arithmeticOverflowError("One or more rows overflow for Add operation.")
```
Comment (gerashegalov) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/arithmetic.scala:58]: let us leave formatting-only changes to dedicated PRs

File: sql-plugin/src/main/spark311/scala/com/nvidia/spark/rapids/shims/ComputeSequenceSizes.scala [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+      withResource(Scalar.fromInt(MAX_ROUNDED_ARRAY_LENGTH)) { maxLen =>
+        withResource(sizeAsLong.lessOrEqualTo(maxLen)) { allValid =>
+          require(isAllValidTrue(allValid),
+            s"Too long sequence found. Should be <= $MAX_ROUNDED_ARRAY_LENGTH")
+        }
+      }
+      // cast to int and return
+      sizeAsLong.castTo(DType.INT32)
+    }
+  }
```
Comment (gerashegalov) [inline sql-plugin/src/main/spark311/scala/com/nvidia/spark/rapids/shims/ComputeSequenceSizes.scala:111]: The bottom portion L85-L111 in 311 and L98-L126 in 351 differ only in the require message let us refactor to minimize shimming

**Review by andygrove — APPROVED** [review]

File: sql-plugin/src/main/spark350/scala/com/nvidia/spark/rapids/shims/api/python/ShimBasePythonRunner.scala
Comment (gerashegalov) [inline sql-plugin/src/main/spark350/scala/com/nvidia/spark/rapids/shims/api/python/ShimBasePythonRunner.scala:1]: This file should be dropped thanks to #9902

**Review by gerashegalov — CHANGES_REQUESTED** [review]

**Comment by razajafri** [comment]
Thanks for taking a look @gerashegalov 

PTAL

**Comment by razajafri** [comment]
build

**Review by gerashegalov — APPROVED** [review]

**Review by NvTimLiu — APPROVED** [review]
LGTM

**Comment by NvTimLiu** [comment]
Will we plan to run nightly integration tests against spark-3.5.1-SNAPSHOT?

**Comment by razajafri** [comment]
> Will we plan to run nightly integration tests against spark-3.5.1-SNAPSHOT?

Yes, we do

**Pull request merged**